src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala

```diff
@@ -25,11 +25,68 @@ trait TreeTraverser {
   }
 
   /**
+   * Use [[TraversalTracing]] for debugging and [[PlainTraversals]] for production
+   */
+  protected type TraversalInstrumentation = PlainTraversals
+
+  protected trait PlainTraversals extends global.Traverser
+
+  protected trait TraversalTracing extends global.Traverser {
+    private var indent = 0
+
+    protected def traceIndividualVisits = false
+
+    override def traverse(t: Tree) = {
+      def doTraverse() = {
+        indent += 2
+        try {
+          super.traverse(t)
+        } finally {
+          indent -= 2
+        }
+      }
+
+      def treeAsString = t.summaryString
+
+      def sourceFileAsString = t.pos match {
+        case null | NoPosition => "<no-source-file>"
+        case properPos => properPos.source.file.name
+      }
+
+      def explainStack = {
+        def stackElemToString(elem: StackTraceElement) = {
+          s"${elem.getFileName}:${elem.getLineNumber}"
+        }
+
+        val stackTrace = Thread.currentThread().getStackTrace.drop(3).take(5)
+        stackTrace.map("  -- " + stackElemToString(_)).mkString("\n")
+      }
+
+      if (indent == 0) {
+        context(s"Starting traversal in $sourceFileAsString at $treeAsString initiated from\n$explainStack\n") {
+          val start = System.currentTimeMillis()
+          doTraverse()
+          val stop = System.currentTimeMillis()
```

Comment:
Well, a lot has been written about this, and there are pros and cons, but I agree with you.